### ISSUE TITLE
string_input_producer possible race while enqueuing

### ISSUE THREAD

**Issue by Utumno** [tensorflow/tensorflow#10256]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows-7-6.1.7601-SP1
- **TensorFlow installed from (source or binary)**: pip
- **TensorFlow version (use command below)**: b'unknown' 1.0.1

Python version `'3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:18:55) [MSC v.1900 64 bit (AMD64)]'
`

### Describe the problem
When I run the script below in pycharm _most of the time_ (hence race) I get 

> ERROR:tensorflow:Exception in QueueRunner: Attempted to use a closed Session.

Running directly from cmd (that's the same command pycharm uses):

```
C:\...>C:\_\Python35\python.exe C:/Users/MrD/.PyCharm2017.1/config/scratches/scratch_40.py
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "BestSplits" device_type: "CPU"') f
or unknown op: BestSplits
...
b'C:\\Dropbox\\eclipse_workspaces\\python\\nn_nielsen\\resources\\tf_records_gap\\img_2013-01-01-00-02.tfrecords'
b'C:\\Dropbox\\eclipse_workspaces\\python\\nn_nielsen\\resources\\tf_records_gap\\img_2013-01-01-00-01.tfrecords'
b'C:\\Dropbox\\eclipse_workspaces\\python\\nn_nielsen\\resources\\tf_records_gap\\img_2013-01-01-00-00.tfrecords'
W c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\kernels\queue_base.cc:294] _0_input_producer: Skipping cancelled enqueue attem
pt with queue not closed
```

### Source code / logs

```python
import os
import tensorflow as tf

data_dir = r"C:\..."
filenames = [os.path.join(data_dir, f) for f in os.listdir(data_dir)]
queue = tf.train.string_input_producer(filenames)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    tf.train.start_queue_runners(sess=sess)
    for _ in range(len(filenames)):
        print(queue.dequeue().eval())
```

Running this as a test inside pycharm (using PyCharm 2017.1.3, Build #PY-171.4424.42) as in:

```python

class TestFileNameQueue(unittest.TestCase):
    def test__get_filename_queue(self):
        data_dir = r"C:\..."
        filenames = [os.path.join(data_dir, f) for f in os.listdir(data_dir)]
        queue = tf.train.string_input_producer(filenames)
        with tf.Session() as sess:
            sess.run(tf.global_variables_initializer())
            tf.train.start_queue_runners(sess=sess)
            for _ in range(len(filenames)):
                print(queue.dequeue().eval())
```

I get a more detailed traceback:

```
C:\_\Python35\python.exe "C:\_\JetBrains\PyCharm 2016.3\helpers\pydev\pydevd.py" --multiproc --save-signatures --qt-support --client 127.0.0.1 --port 4524 --file "C:\_\JetBrains\PyCharm 2016.3\helpers\pycharm\_jb_unittest_runner.py" --target tf_test_sanity.TestFileNameQueue.test__get_filename_queue
Testing started at 11:41 PM ...
warning: Debugger speedups using cython not found. Run '"C:\_\Python35\python.exe" "C:\_\JetBrains\PyCharm 2016.3\helpers\pydev\setup_cython.py" build_ext --inplace' to build.
pydev debugger: process 6148 is connecting

Connected to pydev debugger (build 171.4424.42)
Launching unittests with arguments python -m unittest tf_test_sanity.TestFileNameQueue.test__get_filename_queue in C:\Dropbox\eclipse_workspaces\python\nn_nielsen
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "BestSplits" device_type: "CPU"') for unknown op: BestSplits
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "CountExtremelyRandomStats" device_type: "CPU"') for unknown op: CountExtremelyRandomStats
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "FinishedNodes" device_type: "CPU"') for unknown op: FinishedNodes
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "GrowTree" device_type: "CPU"') for unknown op: GrowTree
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "ReinterpretStringToFloat" device_type: "CPU"') for unknown op: ReinterpretStringToFloat
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "SampleInputs" device_type: "CPU"') for unknown op: SampleInputs
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "ScatterAddNdim" device_type: "CPU"') for unknown op: ScatterAddNdim
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "TopNInsert" device_type: "CPU"') for unknown op: TopNInsert
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "TopNRemove" device_type: "CPU"') for unknown op: TopNRemove
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "TreePredictions" device_type: "CPU"') for unknown op: TreePredictions
E c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\framework\op_kernel.cc:943] OpKernel ('op: "UpdateFertileSlots" device_type: "CPU"') for unknown op: UpdateFertileSlots
ERROR:tensorflow:Exception in QueueRunner: Attempted to use a closed Session.


b'resources/tf_records_no_gap\\img_2013-01-01-00-00.tfrecords'
b'resources/tf_records_no_gap\\img_2013-01-01-00-02.tfrecords'
b'resources/tf_records_no_gap\\img_2013-01-01-00-01.tfrecords'


Ran 1 test in 255.837s

OK
Exception in thread Thread-7:
Traceback (most recent call last):
  File "C:\_\Python35\lib\threading.py", line 914, in _bootstrap_inner
    self.run()
  File "C:\_\Python35\lib\threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "C:\_\Python35\lib\site-packages\tensorflow\python\training\queue_runner_impl.py", line 234, in _run
    sess.run(enqueue_op)
  File "C:\_\Python35\lib\site-packages\tensorflow\python\client\session.py", line 767, in run
    run_metadata_ptr)
  File "C:\_\Python35\lib\site-packages\tensorflow\python\client\session.py", line 903, in _run
    raise RuntimeError('Attempted to use a closed Session.')
RuntimeError: Attempted to use a closed Session.


Process finished with exit code 0
```

**Comment by jart** [comment]
How did you install TensorFlow? Did PyCharm install it?

**Comment by Utumno** [comment]
Nope in cmd:

```    
    C:\Users\MrD>"C:\_\Python35\python.exe" -m pip install --upgrade tensorflow
    Collecting tensorflow
      Downloading tensorflow-1.0.1-cp35-cp35m-win_amd64.whl (14.7MB)
        100% |################################| 14.7MB 37kB/s
    Collecting six>=1.10.0 (from tensorflow)
      Using cached six-1.10.0-py2.py3-none-any.whl
    Collecting wheel>=0.26 (from tensorflow)
      Downloading wheel-0.29.0-py2.py3-none-any.whl (66kB)
        100% |################################| 71kB 1.3MB/s
    Collecting numpy>=1.11.0 (from tensorflow)
      Downloading numpy-1.12.1-cp35-none-win_amd64.whl (7.7MB)
        100% |################################| 7.7MB 73kB/s
    Collecting protobuf>=3.1.0 (from tensorflow)
      Downloading protobuf-3.2.0-py2.py3-none-any.whl (360kB)
        100% |################################| 368kB 875kB/s
    Cache entry deserialization failed, entry ignored
    Collecting setuptools (from protobuf>=3.1.0->tensorflow)
      Downloading setuptools-34.3.2-py2.py3-none-any.whl (389kB)
        100% |################################| 399kB 956kB/s
    Collecting packaging>=16.8 (from setuptools->protobuf>=3.1.0->tensorflow)
      Downloading packaging-16.8-py2.py3-none-any.whl
    Collecting appdirs>=1.4.0 (from setuptools->protobuf>=3.1.0->tensorflow)
      Downloading appdirs-1.4.3-py2.py3-none-any.whl
    Collecting pyparsing (from packaging>=16.8->setuptools->protobuf>=3.1.0->tensorflow)
      Downloading pyparsing-2.2.0-py2.py3-none-any.whl (56kB)
        100% |################################| 61kB 1.9MB/s
    Installing collected packages: six, wheel, numpy, pyparsing, packaging, appdirs, setuptools, protobuf, tensorflow
      Found existing installation: setuptools 20.10.1
        Uninstalling setuptools-20.10.1:
          Successfully uninstalled setuptools-20.10.1
    Successfully installed appdirs-1.4.3 numpy-1.12.1 packaging-16.8 protobuf-3.2.0 pyparsing-2.2.0 setuptools-34.3.2 six-1.10.0 tensorflow-1.0.1 wheel-0.29.0
    You are using pip version 8.1.1, however version 9.0.1 is available.
    You should consider upgrading via the 'python -m pip install --upgrade pip' command.
```

**Comment by jart** [comment]
@mrry I'm confused. Is `tf.contrib` supported on Windows? This is the second such issue I'm noticing today. The reason why I'm confused is because https://github.com/tensorflow/tensorflow/issues/5662 yet it is [excluded](https://github.com/tensorflow/tensorflow/blob/0462416f6498d964310ba05a38a23bd1eabba3bf/tensorflow/python/BUILD#L90) from the bazel build.

**Comment by mrry** [comment]
@jart I don't think this is using any of `tf.contrib`, is it? FWIW, the Windows releases include `tf.contrib` (via the CMake build), but the Bazel build doesn't include it—IIRC because the method of integrating `tf.load_op_library()`-based extensions doesn't currently work in Bazel.

On the original issue, I think this race is "benign", and the error messages are printed in a situation where the queue-runner for `tf.train.string_input_producer()` is terminated by closing the session. I think the errors would be avoided if you add `num_epochs=1` to the`tf.train.string_input_producer()` arguments.

**Comment by Utumno** [comment]
>  I think the errors would be avoided if you add num_epochs=1 to the tf.train.string_input_producer() arguments.

Indeed:

```
import os
import tensorflow as tf

data_dir = r"C:\Dropbox\eclipse_workspaces\python\nn_nielsen\.git\info"
filenames = [os.path.join(data_dir, f) for f in os.listdir(data_dir)]
queue = tf.train.string_input_producer(filenames, capacity=3, num_epochs=1)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    sess.run(tf.local_variables_initializer()) # I had to add this too
    tf.train.start_queue_runners(sess=sess)
    for _ in range(len(filenames)):
        print(queue.dequeue().eval())
```

Note however the error is only printed for small directories (couple items) in my system - for larger ones I get the warning:

> W c:\tf_jenkins\home\workspace\release-win\device\cpu\os\windows\tensorflow\core\kernels\queue_base.cc:294] _0_input_producer: Skipping cancelled enqueue attempt with queue not closed

It's benign as you say but maybe there is something there worth investigating

Thanks!

**Comment by mrry** [comment]
I think the way `tf.train.string_input_producer()` is implemented with background threads and queue runners makes it tricky to avoid this happening. One thing we've been working on recently (and released in TensorFlow 1.2) is a new `Dataset` API that lets you do similar things without having to manage the threads. Here's what your program would look like with the new API:

```python
import os
import tensorflow as tf

data_dir = r"C:\..."
filenames = [os.path.join(data_dir, f) for f in os.listdir(data_dir)]
dataset = tf.contrib.data.Dataset.from_tensor_slices(tf.constant(filenames))
iterator = dataset.make_one_shot_iterator()
next_element = iterator.get_next()

with tf.Session() as sess:
    for _ in range(len(filenames)):
        print(next_element.eval())
```

Note that there would no longer be any need to start queue runners, and the program should exit without errors.

**Comment by Utumno** [comment]
I think the new API does need some polishing (why `tf.constant(filenames)`) but thanks :)
